tools/codegen/api/cpp.py

```diff
@@ -270,16 +274,15 @@ def argument_faithful(
             pin_memory=argument_not_this(a.pin_memory),
         )
     else:
-        return argument(a)
+        return argument(a, is_out_argument)
 
 # NB: this unconditionally groups arguments
+# returns tuple(arguments, out_arguments)
 def group_arguments(
     func: FunctionSchema, *, method: bool
-) -> Sequence[Union[Argument, TensorOptionsArguments, ThisArgument]]:
+) -> Tuple[Sequence[Union[Argument, TensorOptionsArguments, ThisArgument]], Sequence[Argument]]:
```

Comment:
At this point the returned structure should really become a class. The motivation for the imprecise list=of-unions representation was that it left the structure simple and flat, but with this tuple that's no longer the case, and we're going down a slippery slope here (tuple-stapling more and more stuff onto return values without defining a structure for them is like issue #1 when functional programs get too big IME). 

We should bite the bullet and make a `GroupedArguments` data class that models things properly. In the interests of expediency you could probably just do `GroupedArguments` with these two members, but see [this comment thread](https://github.com/pytorch/pytorch/pull/45890#discussion_r503646917) for more.

Reply:
I wanted to show that the refactor in question wasn't so hard; it took me a few hours but now FunctionSchema has grouped arguments in https://github.com/pytorch/pytorch/pull/48195 . I'd like to see this PR reworked on top of it.